pandas/tests/indexes/multi/test_constructor.py

```diff
@@ -83,6 +84,23 @@ def test_constructor_mismatched_codes_levels(idx):
     with pytest.raises(ValueError, match=label_error):
         idx.copy().set_codes([[0, 0, 0, 0], [0, 0]])
 
+    # code value smaller than -1
+    with pytest.raises(ValueError, match=code_value_error):
+        MultiIndex(levels=[['a'], ['b']], codes=[[0, -2], [0, 0]])
+
+
+def test_na_levels():
+    tm.assert_index_equal(
+        MultiIndex(levels=[[np.nan, None, pd.NaT, 128, 2]],
+                   codes=[[0, -1, 1, 2, 3, 4]]),
+        MultiIndex(levels=[[np.nan, None, pd.NaT, 128, 2]],
+                   codes=[[-1, -1, -1, -1, 3, 4]]))
+    tm.assert_index_equal(
+        MultiIndex(levels=[[np.nan, 's', pd.NaT, 128, None]],
```

Comment:
this should also verify if .set_levels and/or .set_codes is called. Can you add tests for each.

Reply:
new test cases added